src/wallet/mod.rs

```diff
@@ -11,196 +11,343 @@ use pbkdf2;
 use sha2::Sha256;
 use shamirsecretsharing::hazmat::{combine_keyshares, create_keyshares};
 use sodiumoxide::randombytes;
-use std::io::{self, Cursor};
+use std::{
+    boxed::Box,
+    fmt,
+    io::{self, Cursor},
+};
 
 pub type Salt = [u8; 8];
 pub type Tag = [u8; 16];
 pub type IV = [u8; 12];
 pub type AESKey = [u8; 32];
 pub type SSSKey = [u8; 32];
 
+#[derive(Clone)]
+pub struct KeyShare(pub(crate) [u8; 33]);
+
 type HmacSha256 = Hmac<Sha256>;
 
-mod basic_wallet;
-pub mod basic {
-    pub use crate::wallet::basic_wallet::*;
-}
-mod sharded_wallet;
-pub mod sharded {
-    pub use crate::wallet::sharded_wallet::*;
+const WALLET_KIND_BASIC: u16 = 0x0001;
+const WALLET_KIND_SHARDED: u16 = 0x0101;
+const WALLET_DEFAULT_ITERATIONS: u32 = 1_000_000;
+
+impl Default for KeyShare {
+    fn default() -> Self {
+        KeyShare([0; 33])
+    }
 }
 
-pub enum Wallet {
-    Basic(basic::Wallet),
-    Sharded(sharded::Wallet),
+impl KeyShare {
+    pub fn to_vec(&self) -> Vec<u8> {
+        self.0.to_vec()
+    }
+
+    pub fn from_slice(slice: &[u8]) -> KeyShare {
+        let mut share = [0u8; 33];
+        share.copy_from_slice(slice);
+        KeyShare(share)
+    }
 }
 
-impl ReadWrite for Wallet {
-    fn read(reader: &mut dyn io::Read) -> Result<Wallet> {
-        let kind = reader.read_u16::<LittleEndian>()?;
-        match kind {
-            0x0001 => {
-                let wallet = basic::Wallet::read(reader)?;
-                Ok(Wallet::Basic(wallet))
-            }
-            0x0101 => {
-                let wallet = sharded::Wallet::read(reader)?;
-                Ok(Wallet::Sharded(wallet))
-            }
-            _ => Err(format!("Invalid wallet type {}", kind).into()),
-        }
+impl fmt::Debug for KeyShare {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "KeyShare({:?})", self.0.to_vec())
     }
+}
 
-    fn write(&self, writer: &mut dyn io::Write) -> Result {
-        match self {
-            Wallet::Basic(wallet) => {
-                writer.write_u16::<LittleEndian>(0x0001)?;
-                wallet.write(writer)
-            }
-            Wallet::Sharded(wallet) => {
-                writer.write_u16::<LittleEndian>(0x0101)?;
-                wallet.write(writer)
-            }
+pub struct Wallet {
+    pub pubkey_bin: PubKeyBin,
+    pub iterations: u32,
+    pub iv: IV,
+    pub salt: Salt,
+    pub tag: Tag,
+    pub encrypted: Vec<u8>,
+    pub format: Box<dyn WalletFormat>,
+}
+
+pub trait WalletFormat {
+    fn wallet_for_keypair(
+        &mut self,
+        keypair: &Keypair,
+        password: &[u8],
+        iterations: u32,
+    ) -> Result<Wallet>;
+    fn keypair_for_wallet(&self, wallet: &Wallet, password: &[u8]) -> Result<Keypair>;
+    fn as_sharded_format(&self) -> Result<ShardedFormat>;
```

Comment:
Is it possible to remove these from the functions (`as_sharded_format` and `absorb_key_shares`) from the Trait since they only apply to ShardedWallets?

Reply:
I couldn't figure out how to convince a handle to a standard WalletFormat trait to take ina new shard. So I had to ask it.. which is what this does?